addon/commands/insert-newLi-command.ts

```diff
@@ -17,261 +17,334 @@ export default class InsertNewLiCommand extends Command {
   }
 
   canExecute(selection: ModelSelection = this.model.selection): boolean {
-    if (!ModelSelection.isWellBehaved(selection)) {
-      throw new MisbehavedSelectionError();
-    }
-    const commonAncestor = selection.getCommonAncestor().parent;
-    if (commonAncestor?.findAncestor(node => ModelNode.isModelElement(node) && (node.type === 'ul' || node.type === 'ol'), true)) {
+    if (selection.isInside(["ul", "ol"]) === "enabled") {
       return true;
+    } else {
+      return false;
     }
-    return false;
   }
-
   execute(): void {
-
     const selection = this.model.selection;
-
-    if (!ModelSelection.isWellBehaved(selection)) {
-      throw new MisbehavedSelectionError();
-    }
     const range = selection.lastRange;
+    if (!range) {
+      throw new Error("couldn't get range");
+    }
+    const startPosition = range.start;
+    const endPosition = range.end;
 
-    const selectionStart = selection.lastRange.start;
-    const selectionEnd = selection.lastRange.end;
-    const startPos = selectionStart.parentOffset;
-    const endPos = selectionEnd.parentOffset;
-    const maximizedRange = range.getMaximizedRange();
+    const isCollapsed = selection.isCollapsed;
 
-    let selectedIterator;
-    if (range.collapsed) {
-      selectedIterator = [];
-      const nodeAfter = range.start.nodeAfter();
-      if (ModelNode.isModelText(nodeAfter)) {
-        selectedIterator.push(nodeAfter);
-      } else if (ModelNode.isModelText(range.start.nodeBefore())){
-        selectedIterator.push(range.start.nodeBefore());
-      }
+    const startParentLi = startPosition.findAncestors(node => ModelNode.isModelElement(node) && node.type === "li")[0];
+    const endParentLi = endPosition.findAncestors(node => ModelNode.isModelElement(node) && node.type === "li")[0];
 
-
-    } else {
-      selectedIterator = new ModelTreeWalker({
-        range: maximizedRange,
-        filter: node => ModelNode.isModelText(node) ? FilterResult.FILTER_ACCEPT : FilterResult.FILTER_SKIP
+    //collapsed selection case
+    if (isCollapsed) {
+      this.insertNewLi(startPosition);
+    }
+    //single li expanded selection case
+    else if (startParentLi == endParentLi) {
```

Comment:
what if both startParentLi and endParentLi are `undefined`

Reply:
perhaps the more general question: should this command block being executed when not already in an li?

Reply:
it won't be, so they "should" never both be undefined

Reply:
@Asergey91 can you add a comment so this assumption is clear when reading the code?

Reply:
oki